source/agora/node/Validator.d

```diff
@@ -66,6 +67,11 @@ public class Validator : FullNode, API
     /// Quorum generator parameters
     protected QuorumParams quorum_params;
 
+    /// The last height at which a quorum shuffle took place.
+    /// If a new block is externalized and `last_shuffle_height`
+    /// is >= `QuorumShuffleCycle` then the quorum will be reshuffled.
+    private Height last_shuffle_height = Height(0);
```

Comment:
I don't think that comment is right, because what you are describing is a target, not a "last". Maybe missing an offset ?

Reply:
No I really am describing when the last shuffle occured.

Reply:
And it is implied that one happened in the genesis block, because that's the first time the quorum "changes".

Reply:
Hum... I see: `last_shuffle_height >= QuorumShuffleCycle`. I thought `QuorumShuffleCycle` was the config option, but it's the target ?

Reply:
I don't understand your terminologies here. Option vs Target.

We want there to be a quorum shuffle at least every X blocks, but they may happen more often if the set of enrollments has changed in the meantime. That's what I commented here https://github.com/bpfkorea/agora/pull/1072#issuecomment-669645266

Reply:
Mondays are hard.

`last_shuffle_height` is height at which the quorum was most recently shuffled;
`target_shuffle_height`, if it existed, would be the height at which the quorum is scheduled to be next shuffled;
`quorum_shuffle_cycle` is the difference between those two values;

Reply:
target would be a misnomer though.

If we have `quorum_shuffle_cycle` set to 10 it might seem a good idea to set `target_shuffle_height` to 10 at the genesis block 0. But if there is a new enrollment at block height 5 then at height 5 we will already shuffle the quorums, so then we would have to shift the target to `15`. 

It seemed unnecessary to me to introduce another state variable when we can just check when was the last time the quorum was shuffled and if the diff is `quorum_shuffle_cycle` then shuffle it again.

Reply:
I did just notice there is a bug in the logic. I was using `>` instead of `>=`.

I'll change the names of these variables.

Reply:
Ah the bug was that `quorum_shuffle_cycle` was not passed to `ConsensusParams`..

This default parameter stuff needs to be removed: https://github.com/bpfkorea/agora/blob/b602254d522f5bc2e5b5e96525c4d61e7c9180ff/source/agora/consensus/data/ConsensusParams.d#L51

It's just going to become a source of bugs.